src/sequence_batch_scheduler.cc

```diff
@@ -110,12 +109,116 @@ SequenceBatchScheduler::Create(
   // runner. This is at least 1 even if the model doesn't support
   // batching.
   const size_t model_batch_size = std::max(1, config.max_batch_size());
-  size_t seq_slot_cnt = model_batch_size;
+  sched->seq_slot_cnt_ = model_batch_size;
   if (config.sequence_batching().has_oldest()) {
-    seq_slot_cnt =
+    sched->seq_slot_cnt_ =
         config.sequence_batching().oldest().max_candidate_sequences();
   }
 
+  // Create batchers.
+  RETURN_IF_ERROR(sched->CreateBatchers(new_instances));
+
+
+  SequenceBatchScheduler* raw = sched.release();
+  // Create background threads that watch for different sequence states.
+  raw->StartBackgroundThreads();
+  scheduler->reset(raw);
+
+  return Status::Success;
+}
+
+Status
+SequenceBatchScheduler::Update(
+    const std::vector<std::shared_ptr<TritonModelInstance>>& added_instances,
+    const std::vector<std::shared_ptr<TritonModelInstance>>& removed_instances)
+{
+  std::lock_guard<std::mutex> lk(mu_);
+
+  // Add the new batchers.
+  RETURN_IF_ERROR(CreateBatchers(added_instances));
+
+  // All sequence slots of the removed instances are pending to be removed.
+  // Track them in 'pending_removal_seq_slots_' and they will be erased once
+  // they become ready.
+  for (auto& instance : removed_instances) {
+    pending_removal_seq_slots_.emplace(
+        instance.get(),
+        std::make_pair(batchers_[instance.get()]->SeqSlotCnt(), instance));
+  }
+
+  // Erase pending removal sequence slots that are already at ready.
+  std::priority_queue<
+      BatcherSequenceSlot, std::vector<BatcherSequenceSlot>,
+      BatcherSequenceSlotCompare>
+      new_ready_batcher_seq_slots;
+  while (!ready_batcher_seq_slots_.empty()) {
+    auto& ready_seq_slot = ready_batcher_seq_slots_.top();
+    if (pending_removal_seq_slots_.find(ready_seq_slot.model_instance_) ==
+        pending_removal_seq_slots_.end()) {
+      // The ready slot is not being removed.
+      new_ready_batcher_seq_slots.push(ready_seq_slot);
+    } else {
+      // The ready slot is being removed, erase the slot.
+      EraseBatcherSequenceSlot(ready_seq_slot);
+    }
+    ready_batcher_seq_slots_.pop();
+  }
+  ready_batcher_seq_slots_.swap(new_ready_batcher_seq_slots);
+
+  return Status::Success;
+}
+
+bool
+SequenceBatchScheduler::EraseBatcherSequenceSlot(
+    const BatcherSequenceSlot& seq_slot)
+{
+  // Find the sequence slot from pending removal sequence slots.
+  auto it = pending_removal_seq_slots_.find(seq_slot.model_instance_);
+  if (it == pending_removal_seq_slots_.end()) {
+    // The sequence slot is not pending removal.
+    return false;
+  }
+
+  LOG_VERBOSE(1) << "Removing slot for batcher "
+                 << seq_slot.model_instance_->Name() << ", slot "
+                 << seq_slot.seq_slot_;
+
+  // Subtract the number of slots, and erase the batcher if no more slots left.
+  if (--it->second.first == 0) {
```

Comment:
Can you assign some names to `auto& descriptive_var = it->first`, `it->second`, and `batcher_it->second`, etc.? It's not easy to scan the following section without going back and checking what each pair consists of.

Reply:
Sure. Updated on commit: [Improve iterator variable naming](https://github.com/triton-inference-server/core/pull/207/commits/f3338a9873683baf899b48b051bc0d37f86add37)